sdk/cosmos/azure_data_cosmos/src/models/patch_operations.rs

```diff
@@ -0,0 +1,323 @@
+use azure_core::Model;
+use serde::{Deserialize, Serialize};
+
+// Cosmos' patch operations are _similar_ to JSON Patch (RFC 6902) in structure, but have different operations.
+
+/// Represents a patch document for Cosmos DB.
+///
+/// A patch document is made up of a collection of patch operations.
+/// Each operation describes how to modify a property of a document.
+/// See <https://learn.microsoft.com/en-us/azure/cosmos-db/partial-document-update> for more information on patch operations.
+///
+/// # Examples
+///
+/// To build up a patch document, use [`PatchDocument::default`] to create an empty document,
+/// then use the various `with_` methods to add operations.
+///
+/// ```rust
+/// # use azure_data_cosmos::models::{PatchDocument, PatchOperation};
+/// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+/// let patch = PatchDocument::default()
+///     .with_add("/color".to_string(), "silver")?
+///     .with_move("/from".to_string(), "/to".to_string())?;
+/// # assert_eq!(patch, PatchDocument {
+/// #     operations: vec![
+/// #         PatchOperation::Add {
+/// #             path: "/color".to_string(),
+/// #             value: serde_json::Value::String("silver".to_string()),
+/// #        },
+/// #        PatchOperation::Move {
+/// #            from: "/from".to_string(),
+/// #            to: "/to".to_string(),
+/// #        },
+/// #    ],
+/// # });
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Model, Default, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct PatchDocument {
+    pub operations: Vec<PatchOperation>,
+}
+
+impl PatchDocument {
+    /// Adds a new "add" operation to the patch document.
+    ///
+    /// See the [type documentation](PatchDocument) for more information on patch operations.
+    ///
+    /// # Arguments
+    /// * `path` - The path to the property to add.
+    /// * `value` - The value to add.
+    pub fn with_add(
+        mut self,
+        path: String,
```

Comment:
Would like some input from you, @heaths . As well as anyone else reviewing this.

Taking paths as `String` does feel a little clunky. We expect users to call this and provide **either** string literals **or** owned `String` values from elsewhere in their application. For methods on `CosmosClient` and friends, we take `&str` because even if they _gave_ us a `String`, we'd be immediately concatenating that into a URL and dropping it, so we might as well be more user-friendly and accept `&str`.

I can see two ways these APIs would be called:

* With a string literal (`&'static str`), because the property path being patched is a constant.
* With a `String` from parsed, read, or received from some other place in the application, because the user, or a config file, or similar, specified the path to be patched.

So, we have these options:

1. Accept `impl Into<String>`. If you pass a literal or any other `&str`, we'll allocate a `String` (perhaps cloning the one the `&str` was borrowed from😬), incurring a hidden allocation. But, we can accept your `String` and move it in to the `PatchOperation` without allocating
2. Accept `&str`. Because we have to own our data (deriving `azure_core::Model` requires `DeserializeOwned`), we have to clone it immediately, incurring a hidden allocation.
3. Accept `Cow<'static, str>`. This allows us to accept string literals and owned Strings without cloning. But it's clunky to write at the call-site: `with_add("/path".into(), ...)`
4. Accept `impl Into<Cow<'static, str>>`. This **would** give us the ability to accept string literals and owned Strings without clunky syntax (`with_add("/path", ...)` and `with_add(my_string, ...)` would both work). However, it brings in more `impl` syntax, which we recently decided to avoid.

For now, I've gone with option 3, because the syntax isn't too much, the intent is clear, and we still get to avoid allocating `String`s for string literals. It does mean string literals have to be passed as `"/path".into()` but unless we go to option 4 (which violates our recent "avoid `impl` types" guidance), I don't think there's a better option.

Reply:
It's not so much that we're trying to avoid `impl`, just reduce it so that the API isn't so "clunky". But we do maintain that it has its uses and the AMQP stack will still use it in places because it makes the API - sort of like what you're running into - more friendly. So, I say go with 1 or 4. I'm leaning toward 1 just because it sounds like you always need to own it anyway, so I don't see a `Cow` offering any advantage, or maybe I misunderstood that particular scenario. Are you saying it's only owned sometimes? If so, go with 4 consistently.

Reply:
I don't think we need to own it. In a lot of cases users are providing static string literals here. We need a reference that lives long enough to go from when the user creates the `PatchDocument` to when they call `patch_item`, which will usually be pretty short.

I contemplated having `PatchOperation` hold a reference to the string but that got really messy with `DeserializeOwned` and `Model`.

I'll go with 4 for now then!